File: server/app/api/routes/knowledge.py
"""知识库管理 API —— Phase 2 新增，不影响 /chat。"""

from fastapi import APIRouter, BackgroundTasks, Depends, File, HTTPException, UploadFile
from sqlalchemy.orm import Session

from app.db import get_db
from app.db.models import Document, KnowledgeBase
from app.schemas.knowledge import DocumentOut, KnowledgeBaseCreate, KnowledgeBaseOut
from app.services.rag.knowledge_service import (
    KnowledgeServiceError,
    create_knowledge_base,
    delete_document,
    delete_knowledge_base,
    get_knowledge_base,
    list_documents,
    list_knowledge_bases,
    run_document_indexing,
    upload_document,
)

router = APIRouter(prefix="/knowledge", tags=["knowledge"])


def _to_kb_out(kb: KnowledgeBase) -> KnowledgeBaseOut:
    return KnowledgeBaseOut(
        id=kb.id,
        name=kb.name,
        description=kb.description,
        document_count=len(kb.documents),
        created_at=kb.created_at,
        updated_at=kb.updated_at,
    )


def _to_doc_out(doc: Document) -> DocumentOut:
    return DocumentOut(
        id=doc.id,
        knowledge_base_id=doc.knowledge_base_id,
        filename=doc.filename,
        file_size=doc.file_size,
        status=doc.status,
        chunk_count=doc.chunk_count,
        error_message=doc.error_message,
        created_at=doc.created_at,
        updated_at=doc.updated_at,
    )


@router.get("/bases", response_model=list[KnowledgeBaseOut])
def get_bases(db: Session = Depends(get_db)) -> list[KnowledgeBaseOut]:
    bases = list_knowledge_bases(db)
    return [_to_kb_out(item) for item in bases]


@router.post("/bases", response_model=KnowledgeBaseOut)
def post_base(payload: KnowledgeBaseCreate, db: Session = Depends(get_db)) -> KnowledgeBaseOut:
    try:
        kb = create_knowledge_base(db, payload.name, payload.description)
    except KnowledgeServiceError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _to_kb_out(kb)


@router.get("/bases/{kb_id}", response_model=KnowledgeBaseOut)
def get_base(kb_id: str, db: Session = Depends(get_db)) -> KnowledgeBaseOut:
    kb = get_knowledge_base(db, kb_id)
    if kb is None:
        raise HTTPException(status_code=404, detail="知识库不存在")
    return _to_kb_out(kb)


@router.delete("/bases/{kb_id}")
def remove_base(kb_id: str, db: Session = Depends(get_db)) -> dict[str, str]:
    try:
        delete_knowledge_base(db, kb_id)
    except KnowledgeServiceError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return {"message": "已删除"}


@router.get("/bases/{kb_id}/documents", response_model=list[DocumentOut])
def get_documents(kb_id: str, db: Session = Depends(get_db)) -> list[DocumentOut]:
    try:
        docs = list_documents(db, kb_id)
    except KnowledgeServiceError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return [_to_doc_out(item) for item in docs]


@router.post("/bases/{kb_id}/documents", response_model=DocumentOut)
async def post_document(
    kb_id: str,
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
) -> DocumentOut:
    if not file.filename:
        raise HTTPException(status_code=400, detail="文件名不能为空")

    content = await file.read()
    try:
        doc = upload_document(db, kb_id, file.filename, content)
    except KnowledgeServiceError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    background_tasks.add_task(run_document_indexing, doc.id)
    return _to_doc_out(doc)


@router.delete("/documents/{doc_id}")
def remove_document(doc_id: str, db: Session = Depends(get_db)) -> dict[str, str]:
    try:
        delete_document(db, doc_id)
    except KnowledgeServiceError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return {"message": "已删除"}

File: web/src/api/knowledge.ts
import type {
  CreateKnowledgeBasePayload,
  KnowledgeBase,
  KnowledgeDocument
} from '@/types/knowledge'

const BASE = '/api/knowledge'

async function parseError(response: Response): Promise<string> {
  const body = await response.json().catch(() => null)
  if (body && typeof body === 'object' && 'detail' in body) {
    const detail = (body as { detail: unknown }).detail
    if (typeof detail === 'string') return detail
  }
  return `请求失败 (${response.status})`
}

export async function fetchKnowledgeBases(): Promise<KnowledgeBase[]> {
  const response = await fetch(`${BASE}/bases`)
  if (!response.ok) throw new Error(await parseError(response))
  return response.json()
}

export async function createKnowledgeBase(
  payload: CreateKnowledgeBasePayload
): Promise<KnowledgeBase> {
  const response = await fetch(`${BASE}/bases`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  })
  if (!response.ok) throw new Error(await parseError(response))
  return response.json()
}

export async function deleteKnowledgeBase(id: string): Promise<void> {
  const response = await fetch(`${BASE}/bases/${id}`, { method: 'DELETE' })
  if (!response.ok) throw new Error(await parseError(response))
}

export async function fetchDocuments(kbId: string): Promise<KnowledgeDocument[]> {
  const response = await fetch(`${BASE}/bases/${kbId}/documents`)
  if (!response.ok) throw new Error(await parseError(response))
  return response.json()
}

export async function uploadDocument(kbId: string, file: File): Promise<KnowledgeDocument> {
  const form = new FormData()
  form.append('file', file)

  const response = await fetch(`${BASE}/bases/${kbId}/documents`, {
    method: 'POST',
    body: form
  })
  if (!response.ok) throw new Error(await parseError(response))
  return response.json()
}

export async function deleteDocument(docId: string): Promise<void> {
  const response = await fetch(`${BASE}/documents/${docId}`, { method: 'DELETE' })
  if (!response.ok) throw new Error(await parseError(response))
}
